File: Src/Icm.Core.Tests/Collections/RepositorySortedCollectionTest.vb
Imports Icm.Collections.Generic
Imports Icm.Collections.Generic.StructKeyClassValue

<TestFixture()>
Public Class RepositorySortedCollectionTest

    Public InitialList As New SortedList(Of Date, String) From {
                {#1/1/2007#, "Item 1a"},
                {#2/1/2007#, "Item 1b"},
                {#1/1/2008#, "Item 2"},
                {#1/1/2009#, "Item 3"},
                {#1/1/2010#, "Item 4"},
                {#1/1/2012#, "Item 5"}
            }

    Public Class FakeRepo
        Inherits MemorySortedCollectionRepository(Of Date, String)

        Public Sub New(ByVal InitialList As SortedList(Of Date, String))
            For Each kvp In InitialList
                List.Add(kvp.Key, kvp.Value)
            Next
        End Sub


    End Class

    <Test()>
    Public Sub Constructor()
        ' ARRANGE
        Dim repo = New FakeRepo(InitialList)

        ' ACT
        Dim sc As New RepositorySortedCollection(Of Date, String)(repo, New YearManager, maxBuckets:=2, totalOrder:=New DateTotalOrder)

        ' ASSERT
        Assert.AreEqual("/", sc.BucketQueue)
        Assert.That(repo.List.SequenceEqual(InitialList))
    End Sub

    <Test()>
    Public Sub ContainsKey_IfThereAreNoDatesInBucket_DoesNotGetBucket()

        ' ARRANGE
        Dim repo = New FakeRepo(InitialList)
        Dim sc As New RepositorySortedCollection(Of Date, String)(repo, New YearManager, maxBuckets:=2, totalOrder:=New DateTotalOrder)
        Dim result As Boolean

        ' ACT
        result = sc.ContainsKey(#1/5/2011#)
        ' ASSERT
        Assert.AreEqual("/", sc.BucketQueue)
        Assert.IsFalse(result)
        Assert.That(repo.List.SequenceEqual(InitialList))
    End Sub

    <Test()>
    Public Sub ContainsKey_IfThereAreDatesInBucket_GetsBucket()
        ' ARRANGE
        Dim repo = New FakeRepo(InitialList)
        Dim sc As New RepositorySortedCollection(Of Date, String)(repo, New YearManager, maxBuckets:=2, totalOrder:=New DateTotalOrder)
        Dim result As Boolean

        ' ACT
        result = sc.ContainsKey(#1/1/2007#)
        ' ASSERT
        Assert.AreEqual("/2007/", sc.BucketQueue)
        Assert.That(result)

    End Sub

    <Test()>
    Public Sub ContainsKey_LastBucketAccessedGetsFirst()
        ' ARRANGE
        Dim repo = New FakeRepo(InitialList)
        Dim sc As New RepositorySortedCollection(Of Date, String)(repo, New YearManager, maxBuckets:=3, totalOrder:=New DateTotalOrder)
        Dim result As Boolean

        ' ACT
        result = sc.ContainsKey(#1/1/2009#) ' -> /2009/
        result = sc.ContainsKey(#1/1/2007#) ' -> /2007/2009/
        result = sc.ContainsKey(#1/5/2011#) ' -> /2007/2009/
        result = sc.ContainsKey(#1/1/2008#) ' -> /2008/2007/2009/
        result = sc.ContainsKey(#1/1/2007#) ' -> /2007/2008/2009/

        ' ASSERT
        Assert.AreEqual("/2007/2008/2009/", sc.BucketQueue)

    End Sub

    <Test()>
    Public Sub ContainsKey_OldestBucketIsThrownAway()
        ' ARRANGE
        Dim repo = New FakeRepo(InitialList)
        Dim sc As New RepositorySortedCollection(Of Date, String)(repo, New YearManager, maxBuckets:=2, totalOrder:=New DateTotalOrder)
        Dim result As Boolean

        ' ACT
        result = sc.ContainsKey(#1/1/2009#)
        result = sc.ContainsKey(#1/1/2007#)

        ' Third year examined causes the oldest (2009) to be expelled.
        result = sc.ContainsKey(#1/1/2008#)
        ' ASSERT
        Assert.AreEqual("/2008/2007/", sc.BucketQueue)
    End Sub

    <Test()>
    Public Sub Next_WhenInSameBucket()
        ' ARRANGE
        Dim repo = New FakeRepo(InitialList)
        Dim sc As New RepositorySortedCollection(Of Date, String)(repo, New YearManager, maxBuckets:=2, totalOrder:=New DateTotalOrder)
        Dim result As Date?
        ' ACT
        result = sc.Next(#1/1/2007#)

        ' ASSERT
        Assert.AreEqual(#2/1/2007#, result)
    End Sub


    <Test()>
    Public Sub Next_WhenInOtherNotLoadedBucket()
        ' ARRANGE
        Dim repo = New FakeRepo(InitialList)
        Dim sc As New RepositorySortedCollection(Of Date, String)(repo, New YearManager, maxBuckets:=2, totalOrder:=New DateTotalOrder)
        Dim result As Date?
        ' ACT
        Assert.AreEqual("/", sc.BucketQueue)
        result = sc.Next(#2/1/2007#)

        ' ASSERT
        Assert.AreEqual(#1/1/2008#, result)
        Assert.AreEqual("/2008/", sc.BucketQueue)
    End Sub


    <Test()>
    Public Sub Next_WhenInOtherLoadedBucket()
        ' ARRANGE
        Dim repo = New FakeRepo(InitialList)
        Dim sc As New RepositorySortedCollection(Of Date, String)(repo, New YearManager, maxBuckets:=2, totalOrder:=New DateTotalOrder)
        Dim result As Date?
        ' ACT
        sc.ContainsKey(#1/1/2008#)

        Assert.AreEqual("/2008/", sc.BucketQueue)
        result = sc.Next(#2/1/2007#)

        ' ASSERT
        Assert.AreEqual(#1/1/2008#, result)
        Assert.AreEqual("/2008/", sc.BucketQueue)
    End Sub

    <Test()>
    Public Sub Next_WhenInOtherNotConsecutiveBucket()
        ' ARRANGE
        Dim repo = New FakeRepo(InitialList)
        Dim sc As New RepositorySortedCollection(Of Date, String)(repo, New YearManager, maxBuckets:=2, totalOrder:=New DateTotalOrder)
        Dim result As Date?
        ' ACT
        result = sc.Next(#1/1/2010#)

        ' ASSERT
        Assert.AreEqual(#1/1/2012#, result)
        Assert.AreEqual("/2012/", sc.BucketQueue)
    End Sub

    <Test()>
    Public Sub Next_WhenDoesNotExist()
        ' ARRANGE
        Dim repo = New FakeRepo(InitialList)
        Dim sc As New RepositorySortedCollection(Of Date, String)(repo, New YearManager, maxBuckets:=2, totalOrder:=New DateTotalOrder)
        Dim result As Date?
        ' ACT
        result = sc.Next(#1/1/2013#)

        ' ASSERT
        Assert.That(result, [Is].Null)

    End Sub
End Class

File: Src/Icm.Core.Tests/Collections/RepositorySortedCollectionTest.cs

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using Icm;
using Icm.Collections.Generic;
using Icm.Collections.Generic.StructKeyClassValue;
using NUnit.Framework;

[TestFixture()]
public class RepositorySortedCollectionTest
{

	public SortedList<System.DateTime, string> InitialList = new SortedList<System.DateTime, string> {
		{
			1/1/2007 12:00:00 AM,
			"Item 1a"
		},
		{
			2/1/2007 12:00:00 AM,
			"Item 1b"
		},
		{
			1/1/2008 12:00:00 AM,
			"Item 2"
		},
		{
			1/1/2009 12:00:00 AM,
			"Item 3"
		},
		{
			1/1/2010 12:00:00 AM,
			"Item 4"
		},
		{
			1/1/2012 12:00:00 AM,
			"Item 5"
		}

	};
	public class FakeRepo : MemorySortedCollectionRepository<System.DateTime, string>
	{

		public FakeRepo(SortedList<System.DateTime, string> InitialList)
		{
			foreach (void kvp_loopVariable in InitialList) {
				kvp = kvp_loopVariable;
				List.Add(kvp.Key, kvp.Value);
			}
		}


	}

	[Test()]
	public void Constructor()
	{
		// ARRANGE
		dynamic repo = new FakeRepo(InitialList);

		// ACT
		RepositorySortedCollection<System.DateTime, string> sc = new RepositorySortedCollection<System.DateTime, string>(repo, new YearManager(), maxBuckets: 2, totalOrder: new DateTotalOrder());

		// ASSERT
		Assert.AreEqual("/", sc.BucketQueue);
		Assert.That(repo.List.SequenceEqual(InitialList));
	}

	[Test()]

	public void ContainsKey_IfThereAreNoDatesInBucket_DoesNotGetBucket()
	{
		// ARRANGE
		dynamic repo = new FakeRepo(InitialList);
		RepositorySortedCollection<System.DateTime, string> sc = new RepositorySortedCollection<System.DateTime, string>(repo, new YearManager(), maxBuckets: 2, totalOrder: new DateTotalOrder());
		bool result = false;

		// ACT
		result = sc.ContainsKey(1/5/2011 12:00:00 AM);
		// ASSERT
		Assert.AreEqual("/", sc.BucketQueue);
		Assert.IsFalse(result);
		Assert.That(repo.List.SequenceEqual(InitialList));
	}

	[Test()]
	public void ContainsKey_IfThereAreDatesInBucket_GetsBucket()
	{
		// ARRANGE
		dynamic repo = new FakeRepo(InitialList);
		RepositorySortedCollection<System.DateTime, string> sc = new RepositorySortedCollection<System.DateTime, string>(repo, new YearManager(), maxBuckets: 2, totalOrder: new DateTotalOrder());
		bool result = false;

		// ACT
		result = sc.ContainsKey(1/1/2007 12:00:00 AM);
		// ASSERT
		Assert.AreEqual("/2007/", sc.BucketQueue);
		Assert.That(result);

	}

	[Test()]
	public void ContainsKey_LastBucketAccessedGetsFirst()
	{
		// ARRANGE
		dynamic repo = new FakeRepo(InitialList);
		RepositorySortedCollection<System.DateTime, string> sc = new RepositorySortedCollection<System.DateTime, string>(repo, new YearManager(), maxBuckets: 3, totalOrder: new DateTotalOrder());
		bool result = false;

		// ACT
		result = sc.ContainsKey(1/1/2009 12:00:00 AM);
		// -> /2009/
		result = sc.ContainsKey(1/1/2007 12:00:00 AM);
		// -> /2007/2009/
		result = sc.ContainsKey(1/5/2011 12:00:00 AM);
		// -> /2007/2009/
		result = sc.ContainsKey(1/1/2008 12:00:00 AM);
		// -> /2008/2007/2009/
		result = sc.ContainsKey(1/1/2007 12:00:00 AM);
		// -> /2007/2008/2009/

		// ASSERT
		Assert.AreEqual("/2007/2008/2009/", sc.BucketQueue);

	}

	[Test()]
	public void ContainsKey_OldestBucketIsThrownAway()
	{
		// ARRANGE
		dynamic repo = new FakeRepo(InitialList);
		RepositorySortedCollection<System.DateTime, string> sc = new RepositorySortedCollection<System.DateTime, string>(repo, new YearManager(), maxBuckets: 2, totalOrder: new DateTotalOrder());
		bool result = false;

		// ACT
		result = sc.ContainsKey(1/1/2009 12:00:00 AM);
		result = sc.ContainsKey(1/1/2007 12:00:00 AM);

		// Third year examined causes the oldest (2009) to be expelled.
		result = sc.ContainsKey(1/1/2008 12:00:00 AM);
		// ASSERT
		Assert.AreEqual("/2008/2007/", sc.BucketQueue);
	}

	[Test()]
	public void Next_WhenInSameBucket()
	{
		// ARRANGE
		dynamic repo = new FakeRepo(InitialList);
		RepositorySortedCollection<System.DateTime, string> sc = new RepositorySortedCollection<System.DateTime, string>(repo, new YearManager(), maxBuckets: 2, totalOrder: new DateTotalOrder());
		System.DateTime? result = default(System.DateTime?);
		// ACT
		result = sc.Next(1/1/2007 12:00:00 AM);

		// ASSERT
		Assert.AreEqual(2/1/2007 12:00:00 AM, result);
	}


	[Test()]
	public void Next_WhenInOtherNotLoadedBucket()
	{
		// ARRANGE
		dynamic repo = new FakeRepo(InitialList);
		RepositorySortedCollection<System.DateTime, string> sc = new RepositorySortedCollection<System.DateTime, string>(repo, new YearManager(), maxBuckets: 2, totalOrder: new DateTotalOrder());
		System.DateTime? result = default(System.DateTime?);
		// ACT
		Assert.AreEqual("/", sc.BucketQueue);
		result = sc.Next(2/1/2007 12:00:00 AM);

		// ASSERT
		Assert.AreEqual(1/1/2008 12:00:00 AM, result);
		Assert.AreEqual("/2008/", sc.BucketQueue);
	}


	[Test()]
	public void Next_WhenInOtherLoadedBucket()
	{
		// ARRANGE
		dynamic repo = new FakeRepo(InitialList);
		RepositorySortedCollection<System.DateTime, string> sc = new RepositorySortedCollection<System.DateTime, string>(repo, new YearManager(), maxBuckets: 2, totalOrder: new DateTotalOrder());
		System.DateTime? result = default(System.DateTime?);
		// ACT
		sc.ContainsKey(1/1/2008 12:00:00 AM);

		Assert.AreEqual("/2008/", sc.BucketQueue);
		result = sc.Next(2/1/2007 12:00:00 AM);

		// ASSERT
		Assert.AreEqual(1/1/2008 12:00:00 AM, result);
		Assert.AreEqual("/2008/", sc.BucketQueue);
	}

	[Test()]
	public void Next_WhenInOtherNotConsecutiveBucket()
	{
		// ARRANGE
		dynamic repo = new FakeRepo(InitialList);
		RepositorySortedCollection<System.DateTime, string> sc = new RepositorySortedCollection<System.DateTime, string>(repo, new YearManager(), maxBuckets: 2, totalOrder: new DateTotalOrder());
		System.DateTime? result = default(System.DateTime?);
		// ACT
		result = sc.Next(1/1/2010 12:00:00 AM);

		// ASSERT
		Assert.AreEqual(1/1/2012 12:00:00 AM, result);
		Assert.AreEqual("/2012/", sc.BucketQueue);
	}

	[Test()]
	public void Next_WhenDoesNotExist()
	{
		// ARRANGE
		dynamic repo = new FakeRepo(InitialList);
		RepositorySortedCollection<System.DateTime, string> sc = new RepositorySortedCollection<System.DateTime, string>(repo, new YearManager(), maxBuckets: 2, totalOrder: new DateTotalOrder());
		System.DateTime? result = default(System.DateTime?);
		// ACT
		result = sc.Next(1/1/2013 12:00:00 AM);

		// ASSERT
		Assert.That(result, Is.Null);

	}
}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
